### PR TITLE
APIM Self host agent extension on arc box

### PR DESCRIPTION
**Pull request by duongthaiha**
This is the PR for the Issue APIM Self Host Extension https://github.com/microsoft/azure_arc/issues/1974 please This to demonstrate ability to run workload on ARC with API still be centralise control on Azure as other Arc Service.
Sorry I dont think I have access to link the original issue in the development.

The script will do the following:

- Deploy Adventure Web API to k3s.
- Set back end of the Web API to AdventureWork SQL Managed Instance.
- Deploy Azure API Management with self-host gateway.
- Deploy Self-host gate way to the k3s
- Configure the connectively from Azure API Management, self-host gateway and Adventure Workload

Copilot Summerisation:
This pull request to the `azure_jumpstart_arcbox` repository includes changes related to the Azure API Management service. The most important changes include adding parameters and resources for the API Management service, setting default values for parameters, adding policies for the service, and adding a PowerShell script for deploying the service and an API.

Main changes:

* <a href="diffhunk://#diff-42baef3c2db31c930e9a4b143c052923c62c838925ebf92087cd95ba15ed3137R1-R99">`azure_jumpstart_arcbox/artifacts/apim/apim.bicep`</a>: Added parameters and resources for the API Management service, including the service itself, an API within the service, a self-hosted gateway, and the gateway's API.
* <a href="diffhunk://#diff-4885492561f8e0b593118b595f2c5014512a48ccd59a97cd20364005ba304981R1-R5">`azure_jumpstart_arcbox/artifacts/apim/apim.bicepparam`</a>: Added a reference to the `apim.bicep` file and set default values for the `publisherEmail` and `publisherName` parameters.
* <a href="diffhunk://#diff-e8e475f1389e9855e88fd7b12f226b6ddd7962cdc43043e917f04040c8d4f282R1-R27">`azure_jumpstart_arcbox/artifacts/apim/adventurework_template.xml`</a>: Added an XML file containing policies for the API Management service, including a `set-backend-service` policy to set the base URL for the backend service.
* <a href="diffhunk://#diff-3ee540ab26f6adc1b9ce5d329481183cb763732cab3ac50027a9dc91fab57debR1-R311">`azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1`</a>: Added a PowerShell script for deploying an AdventureWork API and an API Management service, including setting up environment variables, logging, and deploying to a Kubernetes cluster.

### CODE REVIEW

File: azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1 [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+Write-Host "`n"
+Write-Host "Switch kubectl context to k3s"
+Write-Host "`n"
+
+kubectx arcbox-k3s
+kubectl get nodes
+
+# Build the adventurework api manifest
+Write-Host "`n"
+Write-Host " Build the adventurework api manifest"
```
Comment (chintalavr) [inline azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1:56]:
```suggestion
Write-Host " Build the AdventureWorks API manifest"
```

File: azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1 [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+################################################
+#Switch kubectl context
+Write-Host "`n"
+Write-Host "Switch kubectl context to k3s"
+Write-Host "`n"
+
+kubectx arcbox-k3s
+kubectl get nodes
+
+# Build the adventurework api manifest
```
Comment (chintalavr) [inline azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1:54]:
```suggestion
# Build the AdventureWorks API manifest
```

File: azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1 [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+Write-Host "`n"
+Write-Host " Retrieving SQL MI connection endpoints"
+Write-Host "`n"
+$primaryEndpoint = kubectl get sqlmanagedinstances jumpstart-sql -n arc -o=jsonpath='{.status.endpoints.primary}'
+$primaryEndpoint = $primaryEndpoint.Substring(0, $primaryEndpoint.IndexOf(',')) + ",11433" 
+$sqlConnectionString = "Data Source=$($primaryEndpoint);Initial Catalog=AdventureWorks2019;User ID=$($Env:AZDATA_USERNAME);Password=$($Env:AZDATA_PASSWORD);Encrypt=False;TrustServerCertificate=True"
+$base64ConnectionString = [Convert]::ToBase64String([char[]]$sqlConnectionString)
+
+################################################
+# Deploy AdventureWork API
```
Comment (chintalavr) [inline azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1:44]:
```suggestion
# Deploy AdventureWorks API
```

File: azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1 [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+type: Opaque
+data:
+  AdventureWorkConnection: $($base64ConnectionString)
+"@
+$secretManifest | kubectl apply -f -
+
+
+#Deploy AdvanetureWork API
+Write-Host "`n"
+Write-Host "Deploy AdvanetureWork API"
```
Comment (chintalavr) [inline azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1:74]:
```suggestion
Write-Host "Deploy AdvanetureWorks API"
```

File: azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1 [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+    matchLabels:
+      app: adventurework
+  template:
+    metadata:
+      labels:
+        app: adventurework
+    spec:
+      containers:
+      - name: adventurework
+        image: duongthaiha/adventureworkwebapi:1.0.4
```
Comment (chintalavr) [inline azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1:94]: Do we have this image somewhere in the MS repository? Using a personal library may break in the future.

File: azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1 [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+    targetPort: 80
+  - name: https
+    port: 443
+    targetPort: 443
+
+"@
+$adventureWorkManifest | kubectl apply -f -
+
+################################################
+# Deploy API Management and Self Host Gateway
```
Comment (chintalavr) [inline azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1:127]:
```suggestion
# Deploy API Management and self-hosted gateway
```

File: azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1 [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+################################################
+
+#Update the back end for weather api
```
Comment (chintalavr) [inline azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1:130]:
```suggestion
#Update the back end for weather API
```

File: azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1 [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+Write-Host "`n"
+Write-Host "Update the back end for weather api"
```
Comment (chintalavr) [inline azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1:132]:
```suggestion
Write-Host "Update the back end for weather API"
```

File: azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1 [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+#Deploy API Management and the API
+Write-Host "`n"
+Write-Host "Deploy API Management and the API"
+Write-Host "`n"
+
+$apimDeploymentOutput =  $(az deployment group create --resource-group $Env:resourceGroup --template-file "$Env:ArcBoxDir\apim\apim.bicep" --parameters "$Env:ArcBoxDir\apim\apim.bicepparam") | ConvertFrom-Json
+$apimName = $apimDeploymentOutput.properties.outputs.apiManagementServiceName.value
+
+
+#Get access token to the rest api
```
Comment (chintalavr) [inline azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1:149]:
```suggestion
#Get access token to the REST API
```

File: azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1 [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+Write-Host "`n"
+Write-Host "Get access token to the rest api"
```
Comment (chintalavr) [inline azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1:151]:
```suggestion
Write-Host "Get access token to the REST API"
```

File: azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1 [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+Write-Host "`n"
+
+$access_token = $(az account get-access-token -s $env:subscriptionId --query "accessToken")| ConvertFrom-Json
+
+#Call Rest API to get the selft host gateway token
+Write-Host "`n"
+Write-Host "Call Rest API to get the selft host gateway token"
```
Comment (chintalavr) [inline azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1:158]:
```suggestion
Write-Host "Call REST API to get the self-hosted gateway token"
```

File: azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1 [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+
+
+#Get access token to the rest api
+Write-Host "`n"
+Write-Host "Get access token to the rest api"
+Write-Host "`n"
+
+$access_token = $(az account get-access-token -s $env:subscriptionId --query "accessToken")| ConvertFrom-Json
+
+#Call Rest API to get the selft host gateway token
```
Comment (chintalavr) [inline azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1:156]:
```suggestion
#Call REST API to get the self-hosted gateway token
```

File: azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1 [hunk 13]
```diff
@@ -0,0 +1,311 @@
...
+	expiry: `"${tokenExpire}`",
+	keyType: `"primary`"
+}
+"@
+$generateTokenUrl = 'https://management.azure.com/subscriptions/'+$env:subscriptionId+'/resourceGroups/'+$env:resourceGroup+'/providers/Microsoft.ApiManagement/service/'+$apimName+'/gateways/selfhost/generateToken?api-version=2022-08-01'
+
+$response = Invoke-RestMethod  $generateTokenUrl -Method 'POST' -Headers $headers -Body $body
+
+
+#Create secret for self host gateway
```
Comment (chintalavr) [inline azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1:178]:
```suggestion
#Create a secret for self-hosted gateway
```

File: azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1 [hunk 14]
```diff
@@ -0,0 +1,311 @@
...
+Write-Host "`n"
+Write-Host "Create secret for self host gateway"
```
Comment (chintalavr) [inline azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1:180]:
```suggestion
Write-Host "Create a secret for self-hosted gateway"
```

File: azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1 [hunk 15]
```diff
@@ -0,0 +1,311 @@
...
+Write-Host "`n"
+
+$selfHostToken = "GatewayKey $($response.value)"
+kubectl delete secret selfhost-token
+kubectl create secret generic selfhost-token --from-literal=value="$($selfHostToken)"  --type=Opaque
+
+#Deploy self host agent into k3s
+Write-Host "`n"
+Write-Host "Deploy self host into k3s"
```
Comment (chintalavr) [inline azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1:189]:
```suggestion
Write-Host "Deploy self-hosted gateway into k3s"
```

File: azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1 [hunk 16]
```diff
@@ -0,0 +1,311 @@
...
+  - name: discovery-heartbeat
+    port: 4291
+    targetPort: dc-heartbeat
+    protocol: UDP
+  selector:
+    app: selfhost
+"@    
+$selfHostYaml | kubectl apply -f -
+Write-Host "`n"
+Write-Host "Complete deploy APIM and AdventureWork API"
```
Comment (chintalavr) [inline azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1:309]:
```suggestion
Write-Host "Complete deploy APIM and AdventureWorks API"
```

**Review by chintalavr — CHANGES_REQUESTED** [review]
Please review the suggested changes and update as needed.

File: azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1 (hunk 5 above)
Comment (duongthaiha) [reply azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1:94]: This is a customs image I have created for this workload. If there is a way to uploaded to a shared repo please lets me know. Happy to push it up anywhere the team see fit.

**Review by duongthaiha — COMMENTED** [review]
Include all change in style casing. There is one query regarding the repo. Please lets me know where I can upload the image up please.

File: azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1 (hunk 5 above)
Comment (duongthaiha) [reply azure_jumpstart_arcbox/artifacts/Deploy-APIM.ps1]: I have update the ACR to use jumpstartprod.azurecr.io please

**Review by chintalavr — COMMENTED** [review]
Looks good with the suggested changes.

**Review by chintalavr — APPROVED** [review]
Looks good with the suggested changes.

**Pull request merged**